src/Render.c

```diff
@@ -0,0 +1,276 @@
+                m_vacantPos[i] = true;
+            else
+                m_snowArr[i].age--;
+        }
+    }
+}
+
+static void Render_generateTornado(int frame, ws2812_pixel_t *pixels, int tailHue,
+                     int tailLength, int hue) {
+    for (int x = 0; x < WIDTH; x++) {
+        for (int y = 0; y < HEIGHT; y++) {
+            pixels[x * HEIGHT + y] = Color_hexToRGB(0x000000);
+        }
+    }
+    int l_currentPos = frame % LED_NUMBER;
+    int l_currentX = l_currentPos % WIDTH;
+    int l_currentY = (int)l_currentPos / WIDTH;
+    if (l_currentX % 2)
+        l_currentY = HEIGHT - l_currentY - 1;
+    for (int i = 0; i < tailLength; i++) {
+        int l_tailPos = l_currentPos - 1 - i;
+        if (l_tailPos < 0)
+            l_tailPos = LED_NUMBER + l_tailPos;
+        int l_tailX = l_tailPos % WIDTH;
+        int l_tailY = (int)l_tailPos / WIDTH;
+        if (l_tailX % 2)
+            l_tailY = HEIGHT - l_tailY - 1;
+        int l_currTailPos = (l_tailX * HEIGHT + l_tailY) % LED_NUMBER;
+        hsv_t l_data;
+        l_data.h = tailHue;
+        l_data.s = 1;
+        l_data.v = (float)1 / (i * i * 0.5 + 1);
+        pixels[l_currTailPos] = Color_hsv2rgb(l_data);
+    }
+    hsv_t l_data;
+    l_data.s = 1;
+    l_data.v = 1;
+    l_data.h = hue;
+    pixels[l_currentX * HEIGHT + l_currentY] = Color_hsv2rgb(l_data);
+}
+
+static void Render_shutdown(ws2812_pixel_t *pixels) {
+    for (int x = 0; x < WIDTH; x++) {
+        for (int y = 0; y < HEIGHT; y++) {
+            pixels[x * HEIGHT + y] = Color_hexToRGB(0x000000);
+        }
+    }
+}
+
+static void Render_render(int frame, ws2812_pixel_t *pixels) {
+
+    struct parserData_t *l_params = HttpServer_getCurrentParams();
+
+    int l_hue = l_params->currentHue;
+    int l_mode = l_params->currentMode;
+    int l_speed = l_params->currentSpeed;
+    int l_tailHue = l_params->currentTailHue;
+    int l_snowLife = l_params->currentSnowLife;
+    int l_snowNumber = l_params->currentSnowNumber;
+    int l_tailLength = l_params->currentTailLength;
+    bool l_rainbowAlign = l_params->rainbowAlign;
+    bool l_currentMirror = l_params->currentMirror;
+    bool l_direction = l_params->currentDirection;
+
+    switch (l_mode) {
+    case RAINBOW_MODE:
+        Render_generateRainbow(frame, pixels, l_rainbowAlign, l_speed, l_direction);
+        break;
+    case WAVE_MODE:
+        Render_generateWave(frame, pixels, l_currentMirror, l_speed, l_direction, l_hue);
+        break;
+    case TAPES_MODE:
+        Render_generateTapes(frame, pixels, l_speed, l_hue);
+        break;
+    case SNOW_MODE:
+        Render_generateSnow(pixels, l_hue);
+        break;
+    case TORNADO_MODE:
+        Render_generateTornado(frame, pixels, l_tailHue, l_tailLength, l_hue);
+        break;
+    case DISABLE_MODE:
+        Render_shutdown(pixels);
+        break;
+    default:
+        Render_generateRainbow(frame, pixels, l_rainbowAlign, l_speed, l_direction);
+        break;
```

Comment:
I think that we don't need to copy data to a local variable. We can just put it to function at once. as example:
`Render_generateRainbow(frame, pixels, l_params->rainbowAlign, l_speed, l_direction); =>
 Render_generateRainbow(frame, pixels, l_rainbowAlign, l_params->currentSpeed, l_params->currentDirection);`